drivers/ad5770r/ad5770r.c

```diff
@@ -0,0 +1,650 @@
+ * @param dac_value - value that will be set in the register.
+ * @param channel - the selected channel.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad5770r_set_dac_value(struct ad5770r_dev *dev,
+			      uint16_t dac_value, enum ad5770r_channels channel)
+{
+	uint8_t data[2];
+	int32_t ret;
+
+	data[1] = (uint8_t)AD5770R_CH_DAC_DATA_LSB(dac_value);
+	data[0] = (uint8_t)((dac_value & 0x3FC0) >> 6);
+
+	ret = ad5770r_spi_reg_write_multiple(dev,
+					     AD5770R_CH0_DAC_MSB + 2 * channel,
+					     data, 2);
+
+	return ret;
+};
+
+/**
+ * Set dac value.
+ * @param dev - The device structure.
+ * @param dac_input - value that will be set in the register.
+ * @param channel - the selected channel.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad5770r_set_dac_input(struct ad5770r_dev *dev,
+			      uint16_t dac_input, enum ad5770r_channels channel)
+{
+	uint8_t data[2];
+	int32_t ret;
+
+	data[1] = (uint8_t)AD5770R_CH_DAC_DATA_LSB(dac_input);
+	data[0] = (uint8_t)((dac_input & 0x3FC0) >> 6);
+
+	ret = ad5770r_spi_reg_write_multiple(dev,
+					     AD5770R_CH0_INPUT_MSB + 2 * channel,
+					     data, 2);
+
+	return ret;
+};
+
+/**
+ * Set page mask for dac value and input.
+ * @param dev - The device structure.
+ * @param page_mask - The structure contains the page mask values.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad5770r_set_page_mask(struct ad5770r_dev *dev,
+			      struct dac_page_mask page_mask)
+{
+	int32_t ret;
+	uint8_t data[2];
+
+	data[1] = (uint8_t)AD5770R_CH_DAC_DATA_LSB(page_mask.dac_data_page_mask);
+	data[0] = (uint8_t)((page_mask.dac_data_page_mask & 0x3FC0) >> 6);
+
+	ret = ad5770r_spi_reg_write_multiple(dev,
+					     AD5770R_DAC_PAGE_MASK_MSB,
+					     data, 2);
+
+	data[1] = (uint8_t)AD5770R_CH_DAC_DATA_LSB(page_mask.input_page_mask);
+	data[0] = (uint8_t)((page_mask.input_page_mask & 0x3FC0) >> 6);
+
+	ret = ad5770r_spi_reg_write_multiple(dev,
+					     AD5770R_INPUT_PAGE_MASK_MSB,
+					     data, 2);
+
+	return ret;
+}
+
+/**
+ * Set channel select.
+ * @param dev - The device structure.
+ * @param mask_channel_sel - The array contains channel selection options.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad5770r_set_mask_channel(struct ad5770r_dev *dev,
+				 bool *mask_channel_sel)
+{
+	int32_t ret;
+
+	ret = ad5770r_spi_reg_write(dev,
+				    AD5770R_CH_SELECT,
+				    AD5770R_CH_SELECT_SEL_CH(mask_channel_sel[0], AD5770R_CH0) |
+				    AD5770R_CH_SELECT_SEL_CH(mask_channel_sel[1], AD5770R_CH1) |
+				    AD5770R_CH_SELECT_SEL_CH(mask_channel_sel[2], AD5770R_CH2) |
+				    AD5770R_CH_SELECT_SEL_CH(mask_channel_sel[3], AD5770R_CH3) |
+				    AD5770R_CH_SELECT_SEL_CH(mask_channel_sel[4], AD5770R_CH4) |
+				    AD5770R_CH_SELECT_SEL_CH(mask_channel_sel[5], AD5770R_CH5));
+
+	if (ret)
+		return -1;
+
+	dev->mask_channel_sel = mask_channel_sel;
```

Comment:
also, passing references like this here, should be documented, or you should keep a copy of the data in the driver;
if the user passes a value that is on the stack, the `mask_channel_sel` pointer will point to something invalid at some point in time;

Reply:
We just shouldn't do it. Taking references to other objects and keeping the reference without any object lifetime management is asking for trouble.

Besides the driver never uses this, so maybe just drop it.